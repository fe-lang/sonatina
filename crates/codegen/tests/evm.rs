use dir_test::{dir_test, Fixture};

use hex::ToHex;
use revm::{
    inspector_handle_register,
    interpreter::Interpreter,
    primitives::{
        AccountInfo, Address, Bytecode, Bytes, CancunSpec, Env, ExecutionResult, TransactTo, U256,
    },
    Context, EvmContext, Handler,
};

use sonatina_codegen::{
    critical_edge::CriticalEdgeSplitter,
    domtree::DomTree,
    isa::evm::{opcode::OpCode, EvmBackend},
    liveness::Liveness,
    machinst::{assemble::ObjectLayout, lower::Lower, vcode::VCode},
    stackalloc::SimpleAlloc,
};
use sonatina_ir::{
    cfg::ControlFlowGraph,
    ir_writer::{FuncWriteCtx, IrWrite},
    isa::evm::Evm,
    module::ModuleCtx,
    BlockId, Function,
};
use sonatina_parser::{parse_module, ParsedModule};
use sonatina_triple::{Architecture, OperatingSystem, Vendor};
use std::io::{stderr, Write};

// XXX copied from fe test-utils
#[macro_export]
macro_rules! snap_test {
    ($value:expr, $fixture_path: expr) => {
        let mut settings = insta::Settings::new();
        let fixture_path = ::std::path::Path::new($fixture_path);
        let fixture_dir = fixture_path.parent().unwrap();
        let fixture_name = fixture_path.file_stem().unwrap().to_str().unwrap();

        settings.set_snapshot_path(fixture_dir);
        settings.set_input_file($fixture_path);
        settings.set_prepend_module_to_snapshot(false);
        settings.bind(|| {
            insta::_macro_support::assert_snapshot(
                (insta::_macro_support::AutoName, $value.as_str()).into(),
                std::path::Path::new(env!("CARGO_MANIFEST_DIR")),
                fixture_name,
                module_path!(),
                file!(),
                line!(),
                stringify!($value),
            )
            .unwrap()
        })
    };
}

fn parse_sona(content: &str) -> ParsedModule {
    match parse_module(content) {
        Ok(module) => module,
        Err(errs) => {
            let mut w = stderr();
            for err in errs {
                err.print(&mut w, "[test]", content, true).unwrap();
            }
            panic!("Failed to parse test file. See errors above.")
        }
    }
}

#[dir_test(
    dir: "$CARGO_MANIFEST_DIR/test_files/evm",
    glob: "*.sntn"
)]
fn test_evm(fixture: Fixture<&str>) {
    let parsed = parse_sona(fixture.content());

    let backend = EvmBackend::new(Evm::new(sonatina_triple::TargetTriple {
        architecture: Architecture::Evm,
        vendor: Vendor::Ethereum,
        operating_system: OperatingSystem::Evm(sonatina_triple::EvmVersion::Cancun),
    }));

    let mut v = Vec::new();
    let funcs = parsed
        .debug
        .func_order
        .iter()
        .map(|fref| {
            let (vcode, block_order) = parsed.module.func_store.modify(*fref, |function| {
                let (vcode, block_order) = vcode_for_fn(function, &parsed.module.ctx, &backend);
                let ctx = FuncWriteCtx::with_debug_provider(function, *fref, &parsed.debug);
                vcode.write(&mut v, &ctx).unwrap();
                (vcode, block_order)
            });
            writeln!(v).unwrap();
            (*fref, vcode, block_order)
        })
        .collect();

    write!(&mut v, "\n\n---------------\n\n").unwrap();

    let mut layout = ObjectLayout::new(funcs, 0);
    let mut i = 0;
    while layout.resize(&backend, 0) {
        i += 1;
        println!("resize iteration {i}");
    }
    layout.print(&mut v, &parsed.module, &parsed.debug).unwrap();

    let mut bytecode = Vec::new();
    layout.emit(&backend, &mut bytecode);
    let hex = bytecode.encode_hex::<String>();

    writeln!(&mut v, "\n0x{hex}").unwrap();

    let (res, trace) = run_on_evm(&bytecode);

    writeln!(&mut v, "\n{trace}").unwrap();
    writeln!(&mut v, "\n{res:?}").unwrap();

    snap_test!(String::from_utf8(v).unwrap(), fixture.path());
}

fn run_on_evm(bytecode: &[u8]) -> (ExecutionResult, String) {
    let mut db = revm::InMemoryDB::default();
    let revm_bytecode = Bytecode::new_raw(Bytes::copy_from_slice(bytecode));
    let test_address = Address::repeat_byte(0x12);
    db.insert_account_info(
        test_address,
        AccountInfo {
            balance: U256::ZERO,
            nonce: 0,
            code_hash: revm_bytecode.hash_slow(),
            code: Some(revm_bytecode),
        },
    );

    let mut env = Env::default();
    env.tx.clear();
    env.tx.transact_to = TransactTo::Call(test_address);
    env.tx.data = Bytes::copy_from_slice(&[0, 0, 0, 11, 0, 0, 0, 22]);

    let context = Context::new(
        EvmContext::new_with_env(db, Box::new(env)),
        // TracingInspector::new(TracingInspectorConfig::all());
        TestInspector::new(vec![]),
    );

    let mut evm = revm::Evm::new(context, Handler::mainnet::<CancunSpec>());
    evm = evm
        .modify()
        .append_handler_register(inspector_handle_register)
        .build();

    let res = evm.transact_commit();

    // let mut w = TraceWriter::new(&mut v).use_colors(ColorChoice::Never);
    // w.write_arena(evm.context.external.traces()).unwrap();

    match res.clone() {
        Ok(r) => (r, String::from_utf8(evm.context.external.w).unwrap()),
        Err(e) => panic!("evm failure: {e}"),
    }
}

struct TestInspector<W: Write> {
    w: W,
}

impl<W: Write> TestInspector<W> {
    fn new(w: W) -> TestInspector<W> {
        Self { w }
    }
}

impl<W: Write, DB: revm::Database> revm::Inspector<DB> for TestInspector<W> {
    fn initialize_interp(&mut self, _interp: &mut Interpreter, _context: &mut EvmContext<DB>) {
        writeln!(
            self.w,
            "{:>6}  {:<17} input (stack grows to the right)",
            "pc", "opcode"
        )
        .unwrap();
    }

    fn step(&mut self, interp: &mut Interpreter, _context: &mut EvmContext<DB>) {
        // xxx tentatively writing input stack; clean up
        let pc = interp.program_counter();

        let op = interp.current_opcode();
        let code = unsafe { std::mem::transmute::<u8, revm::interpreter::OpCode>(op) };

        let stack = interp.stack().data();

        write!(
            self.w,
            "{:>6}  {:0>2}  {:<12}  ",
            pc,
            format!("{op:x}"),
            code.info().name(),
        )
        .unwrap();
        let imm_size = code.info().immediate_size() as usize;
        if imm_size > 0 {
            let imm_bytes = interp.bytecode.slice((pc + 1)..(pc + 1 + imm_size));
            writeln!(self.w, "{}", imm_bytes).unwrap();
        } else {
            writeln!(self.w, "{stack:?}").unwrap();
        }
    }

    fn step_end(&mut self, _interp: &mut Interpreter, _context: &mut EvmContext<DB>) {
        // NOTE: annoying revm behavior: `interp.current_opcode()` now returns the next opcode.
    }
}

fn vcode_for_fn(
    function: &mut Function,
    module: &ModuleCtx,
    backend: &EvmBackend,
) -> (VCode<OpCode>, Vec<BlockId>) {
    let mut cfg = ControlFlowGraph::new();
    cfg.compute(function);
    let mut splitter = CriticalEdgeSplitter::new();
    splitter.run(function, &mut cfg);

    let mut liveness = Liveness::new();
    liveness.compute(function, &cfg);
    let mut dom = DomTree::new();
    dom.compute(&cfg);
    let mut alloc = SimpleAlloc::for_function(function, &cfg, &dom, &liveness, 16);
    let lower = Lower::new(module, function);

    match lower.lower(backend, &mut alloc) {
        Ok(vcode) => (vcode, dom.rpo().to_owned()),
        Err(err) => panic!("{:?}", err),
    }
}
