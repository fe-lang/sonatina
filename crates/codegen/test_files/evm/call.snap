---
source: crates/codegen/tests/evm.rs
expression: "String::from_utf8(v).unwrap()"
input_file: test_files/evm/call.sntn
---
// func private %square(v0.i64) -> i64
square:
  block0:
    JUMPDEST
    MUL  // v1.i64 = mul v0 v0;
    SWAP1  // return v1;
    JUMP

// func private %invert(v0.i64) -> i64
invert:
  block0:
    JUMPDEST
    PUSH1 0xff (255)  // v1.i64 = mul v0 -1.i64;
    PUSH0
    SIGNEXTEND
    SWAP1
    MUL
    SWAP1  // return v1;
    JUMP

// func public %main(v0.i8, v1.i64) -> i64
main:
  block0:
    JUMPDEST
    DUP1  // br_table v0 block3 (0.i8 block1) (1.i8 block2);
    PUSH0
    EQ
    PUSH1 block1
    JUMPI
    DUP1
    PUSH1 0x1 (1)
    EQ
    PUSH1 block2
    JUMPI
    PUSH1 block3
    JUMP
  block1:
    JUMPDEST
    PUSH1 `pc + (4)`  // v2.i64 = call %square v1;
    SWAP1
    PUSH1 FuncRef(0)
    JUMP
    JUMPDEST
    SWAP1  // jump block4;
    POP
    PUSH1 block4
    JUMP
  block2:
    JUMPDEST
    PUSH1 `pc + (4)`  // v3.i64 = call %invert v1;
    SWAP1
    PUSH1 FuncRef(1)
    JUMP
    JUMPDEST
    SWAP1  // jump block4;
    POP
    PUSH1 block4
    JUMP
  block3:
    JUMPDEST
    SWAP1  // jump block4;
    POP
  block4:
    JUMPDEST
    SWAP1  // return v4;
    JUMP



---------------

// func private %square(v0.i64) -> i64
square:
  block0:
    0    JUMPDEST
    1    MUL  // v1.i64 = mul v0 v0;
    2    SWAP1  // return v1;
    3    JUMP
// func private %invert(v0.i64) -> i64
invert:
  block0:
    4    JUMPDEST
    5    PUSH1 0xff (255)  // v1.i64 = mul v0 -1.i64;
    7    PUSH0
    8    SIGNEXTEND
    9    SWAP1
   10    MUL
   11    SWAP1  // return v1;
   12    JUMP
// func public %main(v0.i8, v1.i64) -> i64
main:
  block0:
   13    JUMPDEST
   14    DUP1  // br_table v0 block3 (0.i8 block1) (1.i8 block2);
   15    PUSH0
   16    EQ
   17    PUSH1 30 (block1)
   19    JUMPI
   20    DUP1
   21    PUSH1 0x1 (1)
   23    EQ
   24    PUSH1 42 (block2)
   26    JUMPI
   27    PUSH1 55 (block3)
   29    JUMP
  block1:
   30    JUMPDEST
   31    PUSH1 36  // v2.i64 = call %square v1;
   33    SWAP1
   34    PUSH0 0 (FuncRef(0))
   35    JUMP
   36    JUMPDEST
   37    SWAP1  // jump block4;
   38    POP
   39    PUSH1 58 (block4)
   41    JUMP
  block2:
   42    JUMPDEST
   43    PUSH1 49  // v3.i64 = call %invert v1;
   45    SWAP1
   46    PUSH1 4 (FuncRef(1))
   48    JUMP
   49    JUMPDEST
   50    SWAP1  // jump block4;
   51    POP
   52    PUSH1 58 (block4)
   54    JUMP
  block3:
   55    JUMPDEST
   56    SWAP1  // jump block4;
   57    POP
  block4:
   58    JUMPDEST
   59    SWAP1  // return v4;
   60    JUMP


0x5b0290565b60ff5f0b900290565b805f14601e5780600114602a576037565b6024905f565b9050603a565b6031906004565b9050603a565b90505b9056
